fw/sock.c

```diff
@@ -598,62 +668,26 @@ ss_do_close(struct sock *sk, int flags)
 		}
 		tcp_set_state(sk, TCP_CLOSE);
 		tcp_send_active_reset(sk, sk->sk_allocation);
+		goto adjudge_to_death;
 	} else if (tcp_close_state(sk)) {
 		tcp_send_fin(sk);
 	}
 
-adjudge_to_death:
-	sock_hold(sk);
-	sock_orphan(sk);
-
-	/*
-	 * An adoption of release_sock() for our sleep-less tcp_close() version.
-	 *
-	 * SS sockets are processed in softirq only,
-	 * so backlog queue should be empty.
-	 */
-	WARN_ON(sk->sk_backlog.tail);
-	tcp_release_cb(sk);
-	sk->sk_lock.owned = 0;
-	if (waitqueue_active(&sk->sk_lock.wq))
-		wake_up(&sk->sk_lock.wq);
-
-	percpu_counter_inc(sk->sk_prot->orphan_count);
-
-	if (sk->sk_state == TCP_FIN_WAIT2) {
-		const int tmo = tcp_fin_time(sk);
-		if (tmo > TCP_TIMEWAIT_LEN) {
-			inet_csk_reset_keepalive_timer(sk, tmo - TCP_TIMEWAIT_LEN);
-		} else {
-			tcp_time_wait(sk, TCP_FIN_WAIT2, tmo);
-			return;
-		}
+	if (sk_stream_closing(sk)) {
+		sk->sk_lock.owned = 0;
+		return 1;
 	}
-	if (sk->sk_state != TCP_CLOSE) {
-		sk_mem_reclaim(sk);
-		if (tcp_check_oom(sk, 0)) {
-			tcp_set_state(sk, TCP_CLOSE);
-			tcp_send_active_reset(sk, GFP_ATOMIC);
-			__NET_INC_STATS(sock_net(sk),
-					LINUX_MIB_TCPABORTONMEMORY);
-		}
-	}
-	if (sk->sk_state == TCP_CLOSE) {
-		struct request_sock *req;
 
-		req = rcu_dereference_protected(tcp_sk(sk)->fastopen_rsk,
-						lockdep_sock_is_held(sk));
-		if (req)
-			reqsk_fastopen_remove(sk, req, false);
-		if (flags & __SS_F_RST)
-			/*
-			 * Evict all data for transmission since we might never
-			 * have enough window from the malicious/misbehaving client.
-			 * Receive queue is purged in inet_csk_destroy_sock().
-			 */
-			tcp_write_queue_purge(sk);
-		inet_csk_destroy_sock(sk);
-	}
+adjudge_to_death:
+	ss_finish_closing(sk, flags);
+	return 0;
+}
+
+static void
+ss_full_close(struct sock *sk, int flags)
+{
+	if (ss_do_close(sk, flags))
+		ss_finish_closing(sk, flags);
```

Comment:
We need this `ss_finish_closing()` call only if there is some data to send, i.e. we have just the same problem as before (e.g. with TLS alert not sent as in #1975)?